flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/physical/common/CommonTemporalTableJoinRule.scala

```diff
@@ -0,0 +1,134 @@
+ * limitations under the License.
+ */
+package org.apache.flink.table.plan.rules.physical.common
+
+import org.apache.calcite.plan.RelOptRule.{any, operand}
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.core.TableScan
+import org.apache.calcite.rex.{RexNode, RexProgram}
+import org.apache.flink.table.api.TableException
+import org.apache.flink.table.plan.nodes.common.CommonTemporalTableJoin
+import org.apache.flink.table.plan.nodes.logical._
+import org.apache.flink.table.plan.nodes.physical.PhysicalTableSourceScan
+import org.apache.flink.table.sources.{LookupableTableSource, TableSource}
+
+/**
+  * Base implementation for both
+  * [[org.apache.flink.table.plan.nodes.physical.batch.BatchExecTemporalTableJoin]] and
+  * [[org.apache.flink.table.plan.nodes.physical.stream.StreamExecTemporalTableJoin]].
+  */
+trait CommonTemporalTableJoinRule {
+
+  protected def matches(join: FlinkLogicalJoin, tableScan: TableScan): Boolean = {
+    // TODO: shouldn't match temporal table function join
+    if (findTableSource(tableScan).isEmpty) {
+      throw new TableException(
+        "Temporal table join only support join on a LookupableTableSource " +
+          "not on a DataStream or an intermediate query")
+    }
+    // currently temporal table join only support LookupableTableSource
+    isLookupableTableSource(tableScan)
+  }
+
+  protected def findTableSource(relNode: RelNode): Option[TableSource[_]] = {
+    relNode match {
+      case logicalScan: FlinkLogicalTableSourceScan => Some(logicalScan.tableSource)
+      case physicalScan: PhysicalTableSourceScan => Some(physicalScan.tableSource)
+      // TODO: find TableSource in FlinkLogicalIntermediateTableScan
+      case _ => None
+    }
+  }
+
+  protected def isLookupableTableSource(relNode: RelNode): Boolean = {
+    relNode match {
+      case logicalScan: FlinkLogicalTableSourceScan =>
+        logicalScan.tableSource.isInstanceOf[LookupableTableSource[_]]
+      case physicalScan: PhysicalTableSourceScan =>
+        physicalScan.tableSource.isInstanceOf[LookupableTableSource[_]]
+      // TODO: find TableSource in FlinkLogicalIntermediateTableScan
+      case _ => false
+    }
+  }
+
+  protected def transform(
+    join: FlinkLogicalJoin,
+    input: FlinkLogicalRel,
+    tableSource: TableSource[_],
+    period: RexNode,
+    calcProgram: Option[RexProgram]): CommonTemporalTableJoin
+}
+
+abstract class BaseSnapshotOnTableScanRule(description: String)
+  extends RelOptRule(
+    operand(classOf[FlinkLogicalJoin],
+      operand(classOf[FlinkLogicalRel], any()),
+      operand(classOf[FlinkLogicalSnapshot],
+        operand(classOf[TableScan], any()))),
+    description)
+          with CommonTemporalTableJoinRule {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val join = call.rel[FlinkLogicalJoin](0)
+    val tableScan = call.rel[TableScan](3)
+    matches(join, tableScan)
```

Comment:
we should also check if it's snapshotted by proctime?

Reply:
Currently, whether it is snapshotted by proctime or rowtime will both be translated into temporal table join. And will throw exception if it is rowtime when physical translation, because we don't support rowtime temporal join currently.

So I think we don't need to check proctime here.

Reply:
I noticed current TemporalTableJoin is a `SingleRel`, which is not suitable for further extension after we support scanning data into state and provide event time join. So i think it's inappropriate to translate snapshotted with rowtime to `TemporalTableJoin` and then throw exception inside this physical operator.

Reply:
Make sense. 

I'm also rethinking about the physical node name of `TemporalTableJoin` and `TemporalTableFunctionJoin` (i.e. temporal join in Flink). The node names are really confused to users. Actually, the `TemporalTableJoin` is joining a dimension table which is a SingleRel. 

So how about renaming `TemporalTableJoin` to `DimensionTableJoin`, and renaming `TemporalTableFunctionJoin` to `TemporalJoin`. And we can change the translation rule to:

- snapshot on proctime & table source supports `LookupableTableSource` ==> `DimensionTableJoin` 
- snapshot on rowtime & table source ONLY supports `LookupableTableSource` ==> exception
- snapshot on proctime/rowtime & table source supports scanning ==> `TemporalJoin` 

What do you think ?

Reply:
+1 to this idea. How about rename `DimensionTableJoin` to `LookupJoin`, and rename `TemporalJoin` to `TemporalTableJoin`. 

Conceptually `LookupJoin` is very similar with `NestedLoopJoin` but is a `SingleRel`.

Reply:
+1 to `LookupJoin`.

The only concern of `TemporalTableJoin` is that it is actually joining two streams, but there is a `Table` in the name. Do you think it matters?

Reply:
Because i think TemporalTable is more consistent with sql standard. `TemporalJoin` will confuse others about what is temporal? It's not intuitive that this is actually used for temporal table join.  You can consider this name as `TemporalTableJoin ' Join`

Reply:
Ok, let's go forward with `LookupJoin` and `TemporalTableJoin`.